src/sql/src/plan/query.rs

```diff
@@ -977,6 +977,43 @@ fn plan_subquery(
     Ok((expr.project(finishing.project), scope))
 }
 
+/// Canonicalizes `expr` in preparation for a set operation like `UNION` or
+/// `DISTINCT`.
+///
+/// If `restrict_to` is not-None, only the specified columns will be
+/// canonicalized.
+///
+/// At the moment, canonicalization only applies to data of numeric type, so
+/// that e.g. `1.0` and `1.00` are both trimmed to 1 so that they compare as
+/// equivalent.
+fn canonicalize_relation_expr(
+    types: &[ColumnType],
+    expr: HirRelationExpr,
+    restrict_to: Option<&[usize]>,
+) -> HirRelationExpr {
+    let should_consider = |i| match restrict_to {
+        Some(cols) => cols.contains(&i),
+        None => true,
+    };
+    let mut map_exprs = vec![];
+    let mut project_exprs = vec![];
+    for (i, ty) in types.iter().enumerate() {
+        if should_consider(i) && matches!(ty.scalar_type, ScalarType::Numeric { scale: None }) {
```

Comment:
realize this is a prototype, but just want to note for myself:
- This probably wants to match `Numeric` irrespective of scale
- Needs to descend into `List`s to see if their elements are `Numeric`, which would also need canonicalization.